### PR TITLE
Polish ObjectId & Decimal128

### PR DESCRIPTION
**Pull request by leemaguire**
Requires https://github.com/realm/realm-core/pull/3896 for completion

### CODE REVIEW

File: Realm.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -3459,6 +3481,7 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 5DD755E01BE05C19002800DA /* Tests.xcconfig */;
 			buildSettings = {
+				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
```
Comment (jsflax) [inline Realm.xcodeproj/project.pbxproj:3484]: This seems out of place? @tgoyne

File: RealmSwift/ObjectId.swift [hunk 2]
```diff
@@ -81,3 +93,46 @@ extension ObjectId: Encodable {
...
 }
+
+extension ObjectId: Comparable {
+    /// Returns a Boolean value indicating whether the value of the first
+    /// argument is less than that of the second argument.
+    ///
+    /// - Parameters:
+    ///   - lhs: An ObjectId value to compare.
+    ///   - rhs: Another ObjectId value to compare.
+    public static func < (lhs: ObjectId, rhs: ObjectId) -> Bool {
```
Comment (jsflax) [inline RealmSwift/ObjectId.swift:104]: Extra space after operators

File: Realm/RLMDecimal128.h [hunk 3]
```diff
@@ -67,6 +73,49 @@ NS_ASSUME_NONNULL_BEGIN
 
 /// Gets if this Decimal128 represents a NaN value.
 @property (nonatomic, readonly) BOOL isNaN;
+
+/// The magnitude of this RLMDecimal128.
+@property (nonatomic, readonly) RLMDecimal128 *magnitude NS_REFINED_FOR_SWIFT;
+
+/// Replaces this RLMDecimal128 value with its additive inverse.
+- (void)negate NS_REFINED_FOR_SWIFT;
```
Comment (jsflax) [inline Realm/RLMDecimal128.h:81]: Seems to me that `negate` is the only one here that should _not_ be `NS_REFINED_FOR_SWIFT` while the rest should be.

Comment (jsflax) [reply Realm/RLMDecimal128.h:81]: I see now that you were just mirroring the implementation of other swift Number structs, but I still find this a bit odd.

File: Realm.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -3338,6 +3358,7 @@
 					Realm/ObjectStore/external/json,
 				);
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 5.0;
```
Comment (tgoyne) [inline Realm.xcodeproj/project.pbxproj:3361]: This should not be set in the project file (it's set in the xcconfig files).

File: Realm.xcodeproj/project.pbxproj (hunk 1 above)
Comment (tgoyne) [reply Realm.xcodeproj/project.pbxproj:3468]: There shouldn't be any reason that this is needed.

File: RealmSwift/Decimal128.swift [hunk 5]
```diff
@@ -83,10 +83,16 @@ public final class Decimal128: RLMDecimal128, Decodable {
             throw DecodingError.dataCorruptedError(in: container, debugDescription: "Cannot convert value to Decimal128")
         }
     }
-}
 
-extension Decimal128 {
+    /// The mininum value for Decimal128
+    public static var min: Self {
+        minimumDecimalNumber as! Self
```
Comment (jsflax) [inline RealmSwift/Decimal128.swift:89]: Seems like minimum and maximum decimal number should both be refined for swift here.

File: RealmSwift/Decimal128.swift [hunk 6]
```diff
@@ -122,3 +128,191 @@ extension Decimal128: ExpressibleByStringLiteral {
...
+    /// A type that represents the distance between two values.
+    public typealias Stride = Decimal128
+
+    /// Returns the distance from this Decimal128 to the given value, expressed as a
+    /// stride.
+    ///
+    /// - Parameter other: The Decimal128 value to calculate the distance to.
+    /// - Returns: The distance from this value to `other`.
+    public func distance(to other: Decimal128) -> Stride {
+        other-self
```
Comment (jsflax) [inline RealmSwift/Decimal128.swift:292]: Space? Same for the below

File: Realm/RLMDecimal128.h [hunk 7]
```diff
@@ -67,6 +73,49 @@ NS_ASSUME_NONNULL_BEGIN
 
 /// Gets if this Decimal128 represents a NaN value.
 @property (nonatomic, readonly) BOOL isNaN;
+
+/// The magnitude of this RLMDecimal128.
+@property (nonatomic, readonly) RLMDecimal128 *magnitude NS_REFINED_FOR_SWIFT;
+
+/// Replaces this RLMDecimal128 value with its additive inverse.
+- (void)negate;
```
Comment (tgoyne) [inline Realm/RLMDecimal128.h:81]: I don't think we want any mutating functions? NSDecimalNumber doesn't have them and they won't work how users expect them to (i.e. `[foo.decimalProp additionAssignmentWith:num]` won't actually mutate the managed object).

**Review by jsflax — APPROVED** [review]

File: Realm/RLMDecimal128.mm [hunk 8]
```diff
@@ -139,4 +139,82 @@ - (NSString *)stringValue {
...
+}
+
+- (RLMDecimal128 *)decimalNumberByAdding:(RLMDecimal128 *)decimalNumber {
+    auto rhs = RLMObjcToDecimal128(decimalNumber);
+    return [[RLMDecimal128 alloc] initWithDecimal128:_value+rhs];
+}
+
+- (void)additionAssignmentWith:(RLMDecimal128 *)decimalNumber {
+    auto rhs = RLMObjcToDecimal128(decimalNumber);
+    _value+=rhs;
```
Comment (tgoyne) [inline Realm/RLMDecimal128.mm:167]: Spaces around operators for all of these.

File: Realm/Tests/Decimal128Tests.m [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+#import <Realm/RLMDecimal128.h>
+
+@interface Decimal128Tests : RLMTestCase
+@end
+
+@implementation Decimal128Tests
+
+#pragma mark - Initialization
+
+- (void)testDecimal128Initialization {
```
Comment (tgoyne) [inline Realm/Tests/Decimal128Tests.m:29]: The tests should check some values which cannot be represented as doubles.

File: Realm/RLMDecimal128.h (hunk 7 above)
Comment (leemaguire) [reply Realm/RLMDecimal128.h:81]: They are there more-so for the swift implementation. It should be the case that the managed object is mutated

```
- (void)additionAssignmentWith:(RLMDecimal128 *)decimalNumber {
    auto rhs = RLMObjcToDecimal128(decimalNumber);
    _value+=rhs;
}
```

Comment (tgoyne) [reply Realm/RLMDecimal128.h:81]: Modifying the `realm::Decimal128` ivar doesn't write to the Realm (and can't, as it doesn't have any association to the Realm it was read from).

Comment (leemaguire) [reply Realm/RLMDecimal128.h:81]: I've removed the assignment operators but it removes the `SignedNumeric` conformance for `Decimal128`

**Review by jsflax — APPROVED** [review]
Losing conformance to Strideable and SignedNumeric isn't great, but if we choose to add them back in, at least it's an additive change.

**Pull request merged**